dask_match/core.py

```diff
@@ -616,7 +625,245 @@ def optimize_expr(expr):
+                        # of expr nodes that we know we will be
+                        # adding to the local group)
+                        stack.append(dep)
+                    elif dep not in roots and dependencies[dep]:
+                        # Couldn't fuse dep, but we may be able to
+                        # use it as a new root on the next pass
+                        roots.append(dep)
+
+            # Replace fusable sub-group
+            if len(group) > 1:
+                group_deps = []
+                local_names = [_expr._name for _expr in group]
+                for _expr in group:
+                    group_deps += [
+                        operand
+                        for operand in _expr._subgraph_dependencies()
+                        if operand._name not in local_names
+                    ]
+                to_replace = {group[0]: FusedExpr(group, *group_deps)}
+                return replace_nodes(expr, to_replace), not roots
+
+        # Return original expr if no fusable sub-groups were found
+        return expr, True
+
+    while True:
+        original_name = expr._name
+        expr, done = _fusion_pass(expr)
+        if done or expr._name == original_name:
+            break
+
+    return expr
+
+
+@runtime_checkable
+class Fusable(Protocol):
+    """Fusable Expr Protocol
+    An `Expr` with these methods will be treated as "fusable"
+    """
+
+    def _subgraph_dependencies(self):
+        """List of `Expr` operands or `MappedArg`
+        dependencies. Since `_block_subgraph`
+        cannot include mapped arguments explicitly,
+        IO expressions must represent these task
+        arguments as `MappedArg` dependencies.
+        """
+        raise NotImplementedError
+
+    def _block_subgraph(self):
+        """Return the subgraph for an abstract 'block'.
+        Used to initialize a `SubgraphCallabe` object.
+        """
+        raise NotImplementedError
+
+
+class FusedExpr(Expr):
+    @property
+    def exprs(self):
+        return self.operands[0]
+
+    @property
+    def _meta(self):
+        return self.exprs[0]._meta
+
+    def __str__(self):
+        descr = "-".join([expr._name.split("-")[0] for expr in self.exprs])
+        return f"Fused-{descr}"
+
+    @property
+    def _name(self):
+        return f"{str(self)}-{tokenize(self.exprs)}"
+
+    def _divisions(self):
+        return self.exprs[0]._divisions()
+
+    def _subgraph_dependencies(self):
+        return self.operands[1:]
+
+    def _block_subgraph(self):
+        block = {self._name: self.exprs[0]._name}
+        for _expr in self.exprs:
+            for k, v in _expr._block_subgraph().items():
+                block[k] = v
+        return block
+
+    def _layer(self):
+        return _subgraph_callable_layer(
+            self._name,
+            self._block_subgraph(),
+            self._subgraph_dependencies(),
+            self.npartitions,
+        )
+
+
+class MappedArg:
```

Comment:
The term `MappedArg` here is strange to me.  Maybe `Indexable` or something else?   Thoughts on what feels good here?

Reply:
Agree that this feels off. I changed the name to `IndexableArg` for now, but I'm still a bit unhappy with the new name. I'd like to communicate that this object is something that will be eagerly indexed at graph-creation time.